trinity/plugins/builtin/console/plugin.py

```diff
@@ -0,0 +1,91 @@
+from argparse import (
+    Action,
+    ArgumentParser,
+)
+import sys
+
+from trinity.config import (
+    ChainConfig
+)
+from trinity.extensibility import (
+    BaseEvent,
+    BasePlugin,
+    PluginContext,
+    PluginProcess,
+)
+from trinity.extensibility.events import (
+    NetworkProcessReadyEvent,
+    TrinityStartupEvent,
+)
+
+from trinity.plugins.builtin.console.console import (
+    console
+)
+
+
+class ConsolePlugin(BasePlugin):
+
+    def __init__(self) -> None:
+        self.is_attach = False
+        self.is_console = False
+        self.is_vanilla = False
+        self.network_process_ready = False
+        self.chain_config: ChainConfig = None
+
+    @property
+    def name(self) -> str:
+        return "Console"
+
+    @property
+    def process(self) -> PluginProcess:
+        return PluginProcess.MAIN
+
+    def configure_parser(self, arg_parser: ArgumentParser, sub_parser: Action) -> None:
+        console_parser = sub_parser.add_parser(
+            'console', help='run the chain and start the trinity REPL')
+        console_parser.add_argument(
+            '--vanilla-shell',
+            action='store_true',
+            default=False,
+            help='start a native Python shell'
+        )
+
+        attach_parser = sub_parser.add_parser(
+            'attach',
+            help='open an REPL attached to a currently running chain',
+        )
+        attach_parser.add_argument(
+            '--vanilla-shell',
+            action='store_true',
+            default=False,
+            help='start a native Python shell'
+        )
+
+    def handle_event(self, event: BaseEvent) -> None:
+        if isinstance(event, NetworkProcessReadyEvent):
+            self.network_process_ready = True
+
+        if isinstance(event, TrinityStartupEvent):
+            self.is_console = event.args.subcommand == 'console'
+            self.is_attach = event.args.subcommand == 'attach'
+            self.is_vanilla = (hasattr(event.args, "vanilla_shell") and
+                               event.args.vanilla_shell)
+            self.chain_config = event.chain_config
+
+    def should_start(self):
+        return (self.is_console and self.network_process_ready) or self.is_attach
+
+    def start(self, context: PluginContext) -> None:
+        self.run_console(self.chain_config, not self.is_vanilla)
+
+        # If we run `attach` the console will block the node from booting and we want
+        # to shut it down to not continue booting when the user leaves the console
+        if self.is_attach:
+            sys.exit(0)
```

Comment:
It makes sense how we got here. But I'm worried that a pattern where a plugin takes over execution and then exits is a pretty special case. Maybe such a special case that it's not even really a plugin anymore, but a completely different app that's running.

I don't really have a suggestion here other than to maybe add a comment that it's unusual and should be discouraged for plugins to block indefinitely on `start()`, and also discouraged to `exit()`. If two plugins tried to do that at the same time, they would be racing to go first.

Reply:
In the model I proposed (in another comment) we would get rid of the `start` and `should_start` mechanisms entirely in favor of the plugin having a formal route to take over the whole process. (this however requires we drop the `console` command which I'm generally in favor of because I think running both the node and the console in the same process is weird.

Reply:
It *is* weird. Although I do find myself doing it fairly regularly with geth. I guess it's fairly easy to simulate with something like:

```sh
trinity >>info.log &
trinity attach
pkill trinity
```

Probably the thing I would miss the most is making a node with a custom IPC location and being able to start it up with a console without having to retype the IPC location on the second command.

Reply:
I have found that `console` has so much logging noise that it is borderline un-usable.  IMHO it is a weird carryover from geth that shouldn't exist (combined REPL with a bunch of stdout messages clobbering the UI)

Reply:
> I have found that console has so much logging noise that it is borderline un-usabl

Yeah, I basically always do this:

```
geth ... console 2>geth.log
```